### PR TITLE
Data augmentation

### LINKED ISSUE #4: Data Augmentor for a batch
Paradigm: Frame by Frame Entry
Model Specific or General: General
Paradigm Specific: No

The batch generator in issue #3 will generate a batch of images, these images are to taken by the augmentor and apply various augmentations (details on project board).

Link to project board: https://github.com/andy6975/Memoir/projects/1#card-35889770

### PR DESCRIPTION
**Pull request by akshita0208**

### CODE REVIEW

File: Data/data_augmentation.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+        # random transformation to apply for a single image
+        key = random.choice(list(available_transformations))
+        transformed_image = available_transformations[key](image_to_transform)
+        num_transformations += 1
+
+new_file_path = '%s/augmented_image_%s.jpg' % (folder_path, num_generated_files)
+
+# write image to the disk
+io.imsave(new_file_path, transformed_image)
+num_generated_files += 1
```
Comment (shivam7569) [inline Data/data_augmentation.py]: Lines 108-112 should be inside the loop...

File: Data/data_augmentation.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    'affine': vc_affine,
+    'crop': vc_crop,
+    'contrast': vc_contrast
+}
+
+
+folder_path = '/Desktop/videocolor'
+
+#no. of images we want 
+num_files_desired = 10
```
Comment (shivam7569) [inline Data/data_augmentation.py]: Make this user input, better if it is a function parameter.

File: Data/data_augmentation.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    'noise': vc_noise,
+    'brightness': vc_multi,
+    'affine': vc_affine,
+    'crop': vc_crop,
+    'contrast': vc_contrast
+}
+
+
+folder_path = '/Desktop/videocolor'
+
```
Comment (shivam7569) [inline Data/data_augmentation.py:132]: The folder_path: I think there is no need for a path to images. Data augmentor is to be called on the batch of images received from batch_generator, something like this.

image_names = image_names_generator(v_type='Animated', series='All')
batch = batch_generator(image_names, batch_size=100, image_size=(100, 300))
augmented_batch = data_augmentor(batch, number_of_images_to_augment)

augmented_batch will be directly fed to the network, so it must have both the types of images, augmented and original. Basically it should return a N-d array.

File: Data/data_augmentation.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+#no. of images we want 
+num_files_desired = 10 
+
+# find all files paths from the folder
+images = [os.path.join(folder_path, f) for f in os.listdir(folder_path) if os.path.isfile(os.path.join(folder_path, f))]
```
Comment (shivam7569) [inline Data/data_augmentation.py:137]: No need for this. As written in the previous comments:

image_names = image_names_generator(v_type='Animated', series='All')
...
...

the image_names variable has the names of all the images concerned. Just pass this as a function paramter (if you really need to use this information) to augmentor and read a random one. There is no need to repeat the tasks, right!? :)

File: Data/data_augmentation.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+    if is_random:
+        gamma = np.random.uniform(1-gamma, 1+gamma)
+    return exposure.adjust_gamma(image_array, gamma, gain)
+   
+
+#crop
+
+def vc_crop(image_array: ndarray):
+    """Pad image by n_pixels on each size, then take random crop of same
+    original size.
```
Comment (shivam7569) [inline Data/data_augmentation.py:43]: What is n_pixels?

File: Data/data_augmentation.py [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+# dictionary of the transformations we defined earlier
+available_transformations = {
+    'rotate': vc_rotation,
+    'noise': vc_noise,
+    'brightness': vc_multi,
+    'affine': vc_affine,
+    'crop': vc_crop,
+    'contrast': vc_contrast
+}
+
```
Comment (shivam7569) [inline Data/data_augmentation.py:130]: Kindly add filtering techniques too. Here, this might be useful as a reference:

https://hub.packtpub.com/image-filtering-techniques-opencv/

The above link is not exhaustive, and I suggest you find more such augmentation techniques which affects the training. It's never just the model which is responsible for results ;)

[Saturation, resizing, scaling, flipping and perspective transformation are also missing.]

https://medium.com/@prasad.pai/data-augmentation-techniques-in-cnn-using-tensorflow-371ae43d5be9#2f4a

File: Data/data_augmentation.py [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+# find all files paths from the folder
+images = [os.path.join(folder_path, f) for f in os.listdir(folder_path) if os.path.isfile(os.path.join(folder_path, f))]
+
+num_generated_files = 0
+while num_generated_files <= num_files_desired:
+    # random image from the folder
+    image_path = random.choice(images)
+    # read image as an two dimensional array of pixels
+    image_to_transform = sk.io.imread(image_path)
```
Comment (shivam7569) [inline Data/data_augmentation.py:144]: No need to read the images...just pick a random one from the batch of images returned by batch_generator.

**Review by shivam7569 — CHANGES_REQUESTED** [review]
Hi! Thanks Akshita for the PR and amazing work with the augmentor. Here are some changes/additions suggested for the whole code in general (the details are commented in the code):

* First the obvious, kindly convert the code to PEP8 format (the default for the project).
* A little bit more efforts on the documentation would be great. Apologies for the trouble...
* Mention the use of unconventional bits in the code, like the function annotations.

**Review by shivam7569 — APPROVED** [review]

**Review by shivam7569 — APPROVED** [review]

File: Data/data_augmentation.py (hunk 5 above)
Comment (akshita0208) [reply Data/data_augmentation.py:43]: if we are not assuming , its a square image , we are padding it with some pixels.


or we can use this to change the height and width,
from skimage.util import crop


# crop_width{sequence, int}: Number of values to remove from the edges of each axis. 
# ((before_1, after_1), … (before_N, after_N)) specifies unique crop widths at the 
# start and end of each axis. ((before, after),) specifies a fixed start and end 
# crop for every axis. (n,) or n for integer n is a shortcut for before = after = n 
# for all axes.
B = crop(A, ((50, 100), (50, 50), (0,0)), copy=False)

print(A.shape, B.shape)
# (220, 220, 3) (70, 120, 3)

File: Data/data_augmentation.py (hunk 6 above)
Comment (akshita0208) [reply Data/data_augmentation.py:130]: added them

**Review by shivam7569 — APPROVED** [review]
